### PR TITLE
Avoid redundant dataset loading

### PR DESCRIPTION
**Pull request by yngve-sk**

### CODE REVIEW

File: src/ert/storage/local_experiment.py [hunk 1]
```diff
@@ -335,6 +335,43 @@ def response_key_to_response_type(self) -> Dict[str, str]:
...
         for config in self.response_configuration.values():
             for key in config.keys:
+                if key == "*":
+                    continue
+
                 mapping[key] = config.response_type
 
         return mapping
+
+    def update_response_keys(
```
Comment (oyvindeide) [inline src/ert/storage/local_experiment.py:345]: Suggest prefacing this with `_` as it should not be used externally.

Comment (oyvindeide) [reply src/ert/storage/local_experiment.py:345]: Could you also write a short doc-string explaining why we need to do this? Was not obvious to me before I got it explained.

File: src/ert/storage/local_experiment.py [hunk 2]
```diff
@@ -335,6 +335,43 @@ def response_key_to_response_type(self) -> Dict[str, str]:
...
+        if response_type not in responses_configuration:
+            raise KeyError(
+                f"Response type {response_type} does not exist in current responses.json"
+            )
+
+        config = responses_configuration[response_type]
+
+        diff = set(new_response_keys) - set(config.keys)
+
+        if len(diff) > 0:
```
Comment (oyvindeide) [inline src/ert/storage/local_experiment.py:363]: Nitpick, but could also write:
```suggestion
        if diff:
```

and perhaps rename the variable for clarity? Think I would have used `new_response_keys` for this, and the variable you already have named that would be more like: `response_keys`?

Then it becomes:
```python
if new_response_keys:
```
which to me is more readable.

File: src/ert/storage/local_experiment.py [hunk 3]
```diff
@@ -335,6 +335,43 @@ def response_key_to_response_type(self) -> Dict[str, str]:
         mapping = {}
         for config in self.response_configuration.values():
             for key in config.keys:
+                if key == "*":
+                    continue
```
Comment (oyvindeide) [inline src/ert/storage/local_experiment.py:339]: Any way we can avoid this, and rather replace it once we know what the actual responses are called? My main concern is that we are propagating the wilcard to far into the logic.

**Review by oyvindeide — COMMENTED** [review]
Looks like a good change, I have some comments though, but a lot of them are nitpicks. Is: [(squash) Add mutation of response keys to stateful storage test](https://github.com/equinor/ert/pull/9016/commits/08cf168f14a5ad9e8421603e57cf110a33e5a3c7) meant to be squashed into the first commit? If not I think that commit deserves a test. Suggest adding commits as `--fixup` so it is easier to tell which commit they are intended to be merged with.

File: src/ert/storage/local_experiment.py (hunk 3 above)
Comment (yngve-sk) [reply src/ert/storage/local_experiment.py:339]: Hmm I think we could maybe get rid of it yes, should be some property of the respones config I'd say. If we get rid of it now, then the information that we accept wildcards is no longer there. Does it make sense that we only account of the "*" on the "read from runpath to storage step" and ignore it afterwards, and always read all keys that make it into storage?

Comment (oyvindeide) [reply src/ert/storage/local_experiment.py:339]: I think that makes sense, I guess that will already work with the changes you have here? As you would update the keys the first time we save something from runpath?

File: src/ert/dark_storage/common.py [hunk 4]
```diff
@@ -168,30 +201,6 @@ def data_for_key(
             return data.astype(float)
         except ValueError:
             return data
-    if key in gen_data_keys(ensemble):
```
[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/ert/dark_storage/common.py:171]: The current use of the `key@N` notation prevents this at the moment, but I renamed it to `gen_data_display_keys` to make it more clear

File: src/ert/dark_storage/common.py [hunk 5]
```diff
@@ -101,35 +101,68 @@ def data_for_key(
...
+                )
+                mask = ensemble.get_realization_mask_with_responses(response_key)
+                realizations = np.where(mask)[0]
+                data = ensemble.load_responses(response_key, tuple(realizations))
+            except ValueError as err:
+                print(f"Could not load response {key}: {err}")
+                return pd.DataFrame()
+
+            try:
+                vals = data.filter(polars.col("report_step").eq(report_step))
```
[the remark this answers is not in the document]
Comment (oyvindeide) [reply src/ert/dark_storage/common.py:163]: Did not see the indent

File: src/ert/dark_storage/common.py [hunk 6]
```diff
@@ -101,35 +101,68 @@ def data_for_key(
...
+            data = df.unstack(level="Date")
+            data.columns = data.columns.droplevel(0)
+            try:
+                return data.astype(float)
+            except ValueError:
+                return data
+
+        if response_type == "gen_data":
+            try:
+                # Call below will ValueError if key ends with H,
```
[the remark this answers is not in the document]
Comment (oyvindeide) [reply src/ert/dark_storage/common.py:150]: This sounds like a potential bug waiting to happen in PlotApi, will you create an issue to fix it?

File: src/ert/storage/local_experiment.py (hunk 1 above)
Comment (yngve-sk) [reply src/ert/storage/local_experiment.py]: https://github.com/equinor/ert/issues/9059

File: src/ert/dark_storage/common.py [hunk 7]
```diff
@@ -101,35 +101,68 @@ def data_for_key(
...
+                pivoted = vals.drop("response_key", "report_step").pivot(
+                    on="index", values="values"
+                )
+                data = pivoted.to_pandas().set_index("realization")
+                data.columns = data.columns.astype(int)
+                data.columns.name = "axis"
+                try:
+                    return data.astype(float)
+                except ValueError:
+                    return data
```
[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/ert/dark_storage/common.py:173]: Seems it was there in '22 at least:
```
def data_for_key(res: LibresFacade, case, key, realization_index=None) -> pd.DataFrame:
    """Returns a pandas DataFrame with the datapoints for a given key for a
    given case. The row index is the realization number, and the columns are an
    index over the indexes/dates"""

    if key.split(":")[0][-1] == "H":
        return res.history_data(key, case).T

    if key.startswith("LOG10_"):
        key = key[6:]

    if res.is_summary_key(key):
        data = res.gather_summary_data(case, key, realization_index).T
    elif res.is_gen_kw_key(key):
        data = res.gather_gen_kw_data(case, key, realization_index)
        data.columns = pd.Index([0])
    elif res.is_gen_data_key(key):
        data = res.gather_gen_data_data(case, key, realization_index).T
    else:
        raise ValueError(f"no such key {key}")

    try:
        return data.astype(float)
    except ValueError:
        return data
```

File: src/ert/dark_storage/common.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/ert/dark_storage/common.py:150]: https://github.com/equinor/ert/issues/9059

File: src/ert/dark_storage/common.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/ert/dark_storage/common.py:173]: Also not covered by tests, the codecov shows that quite some of the edge cases here are not hit by tests:
https://app.codecov.io/gh/equinor/ert/pull/9016/blob/src/ert/dark_storage/common.py#L146

File: src/ert/storage/local_experiment.py [hunk 8]
```diff
@@ -335,6 +335,49 @@ def response_key_to_response_type(self) -> Dict[str, str]:
...
+            raise KeyError(
+                f"Response type {response_type} does not exist in current responses.json"
+            )
+
+        config = responses_configuration[response_type]
+
+        new_response_keys = set(response_keys) - set(config.keys)
+
+        if new_response_keys:
+            config.keys = sorted(set(config.keys).union(set(response_keys)))
```
Comment (oyvindeide) [inline src/ert/storage/local_experiment.py:370]: After this PR, this becomes that way of reading the response keys? But we still have `get_summary_keyset` which reads the whole dataset, I guess this should use this as well?

File: src/ert/storage/local_experiment.py [hunk 9]
```diff
@@ -334,7 +334,67 @@ def observation_keys(self) -> List[str]:
...
+    def _has_finalized_response_keys(self, response_type: str) -> bool:
+        responses_configuration = self.response_configuration
+        if response_type not in responses_configuration:
+            raise KeyError(
+                f"Response type {response_type} does not exist in current responses.json"
+            )
+
+        return responses_configuration[response_type].has_finalized_keys
+
+    def _update_response_keys(
```
Comment (oyvindeide) [inline src/ert/storage/local_experiment.py:366]: What happens to wildcards in `keys` here?

File: tests/ert/unit_tests/storage/snapshots/test_storage_migration/test_that_storage_matches/responses [hunk 10]
```diff
@@ -1 +1 @@
-{'gen_data': GenDataConfig(name='gen_data', input_files=['gen%d.txt'], keys=['GEN'], report_steps_list=[[1]]), 'summary': SummaryConfig(name='summary', input_files=['CASE'], keys=['FOPR', 'RWPR'], refcase={})}
+{'gen_data': GenDataConfig(name='gen_data', input_files=['gen%d.txt'], keys=['GEN'], has_finalized_keys=True, report_steps_list=[[1]]), 'summary': SummaryConfig(name='summary', input_files=['CASE'], keys=['FOPR', 'RWPR'], has_finalized_keys=False, refcase={})}
```
Comment (oyvindeide) [inline tests/ert/unit_tests/storage/snapshots/test_storage_migration/test_that_storage_matches/responses:1]: In this case, `SummaryConfig` probably has finalized keys? Should we have some sort of logic for that? For example if there are no `*` in `keys` -> `has_finalized_keys` == True?

**Review by oyvindeide — COMMENTED** [review]
Looks good, I just have some minor comments

File: src/ert/dark_storage/common.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (yngve-sk) [reply src/ert/dark_storage/common.py:173]: Done: https://github.com/equinor/ert/issues/9203

File: src/ert/storage/local_experiment.py (hunk 9 above)
Comment (yngve-sk) [reply src/ert/storage/local_experiment.py:366]: They will "disappear", and be replace with the actual first saved keys. So we expect no wildcards whenever `.has_finalized_keys` is `True`

File: tests/ert/unit_tests/storage/snapshots/test_storage_migration/test_that_storage_matches/responses (hunk 10 above)
Comment (yngve-sk) [reply tests/ert/unit_tests/storage/snapshots/test_storage_migration/test_that_storage_matches/responses:1]: Hmm, the way it is set up now, the `has_finalized_keys` is set initially in the response config implementation, so for `SummaryConfig` it always starts as `False`. Then when some responses are saved, those keys are internalized and the flag is switched to `True`. So the logic is there, but the flag is set explicitly when keys are saved explicitly. 

Corner case of having it like this is if the flag is defaulted to `True` and somehow we also have wildcards (would not work), or if we get different keys saved from different responses (should also not happen)

Comment (oyvindeide) [reply tests/ert/unit_tests/storage/snapshots/test_storage_migration/test_that_storage_matches/responses:1]: Would keep the default to `False` I think, but perhaps have it changed in a `__post_init__`, but might be overkill.

Comment (yngve-sk) [reply tests/ert/unit_tests/storage/snapshots/test_storage_migration/test_that_storage_matches/responses:1]: I think we can adjust it later and do something more in `__post_init__` if it becomes necessary, but for now I think we can have it defaulted to `False` in the `ResponseConfig` and `GenDataConfig`, and `True` in `SummaryConfig`, which is in line with the current behavior.

**Review by oyvindeide — APPROVED** [review]
LGTM!
